### PR TITLE
Added support for numpy.newaxis

### PR DESCRIPTION
**Pull request by kc611**
As titled, this PR adds support for `numpy.newaxis` (which is simply an alias for `None`) for indexing NumPy arrays in Numba JIT functions.

```python
import numpy as np
import numba

@numba.njit
def foo(x):
    # Basic Indexing
    return x[:, np.newaxis]

@numba.njit
def foo_fancy(x):
    # Fancy Indexing
    return x[:, np.newaxis, np.array([1,2,1])]

x = np.random.randint(0, 10, (4,5,6))

print(foo(x).shape == foo.py_func(x).shape)
# True

print(foo_fancy(x).shape == foo_fancy.py_func(x).shape)
# True
```

### CODE REVIEW

File: numba/tests/test_fancy_indexing.py [hunk 1]
```diff
@@ -255,6 +255,19 @@ def check(arr, ind):
...
+            return a[idx, np.newaxis, :]
+        
+        arr = np.random.randint(0, 10, (4,5,6))
+        idx = np.array([1,2,1])
+        py_func = np_new_axis
+        cfunc = njit(py_func)
+
+        expected = py_func(arr, idx)
+        got = cfunc(arr, idx)
+        np.testing.assert_equal(expected, got)
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:270]: Think it'd be a good idea to include more tests:
* Multiple `np.newaxis` in a single indexing call.
* `np.newaxis` with other involved indexes.
* `np.newaxis` in a `__setitem__` context.
* Use of `None` in place of `np.newaxis`.

File: numba/np/arrayobj.py [hunk 2]
```diff
@@ -981,7 +990,11 @@ def prepare(self):
         for i in self.indexers:
             i.prepare()
         # Compute the resulting shape
-        self.indexers_shape = sum([i.get_shape() for i in self.indexers], ())
+        shape_list = [i.get_shape() for i in self.indexers]
+        one = self.context.get_constant(types.intp, 1)
+        for i in self.newaxes:
+            shape_list.insert(i, (one,))
+        self.indexers_shape = sum(shape_list, ())
```
Comment (stuartarchibald) [inline numba/np/arrayobj.py:997]: Is this correct? I think this might trigger an issue present here:
```python
from numba import njit
import numpy as np

@njit
def foo(a):
    return a[np.newaxis, :, np.newaxis,  ..., np.newaxis]

def gen(n):
    return np.arange(np.prod(range(1, n + 1))).reshape((*range(n, 0, -1)))

x = gen(4)
got, expected = foo(x), foo.py_func(x)
np.testing.assert_allclose(got, expected)
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. I've given this an initial review, I think that this makes most of the common uses for `np.newaxis` work but there's some edge cases that need looking at. Adding some more tests would probably be a good idea given how indexing is always an involved and nuanced topic! Thanks again!

File: numba/np/arrayobj.py (hunk 2 above)
Comment (kc611) [reply numba/np/arrayobj.py:1002]: Resolved in [be5d0cc](https://github.com/numba/numba/pull/8513/commits/be5d0ccbbff8b33615fc6e2a6eb88c104fc5dbf2)

File: numba/tests/test_fancy_indexing.py (hunk 1 above)
Comment (kc611) [reply numba/tests/test_fancy_indexing.py]: Resolved in [be5d0cc](https://github.com/numba/numba/pull/8513/commits/be5d0ccbbff8b33615fc6e2a6eb88c104fc5dbf2)

File: numba/tests/test_array_manipulation.py [hunk 3]
```diff
@@ -633,23 +629,40 @@ def test_squeeze_array_npm(self):
...
+        def np_new_axis_getitem(a, idx):
+            return a[idx]
+
+        def np_new_axis_setitem(a, idx, item):
+            a[idx] = item
+            return a
+
+        a = np.arange(4*5*6*7).reshape((4, 5, 6, 7))
+        idx_cases = [
+            (np.newaxis, slice(None)),
```
Comment (stuartarchibald) [inline numba/tests/test_array_manipulation.py:642]: Perhaps add some use cases where `np.newaxis` isn't the first item?

File: numba/tests/test_fancy_indexing.py [hunk 4]
```diff
@@ -255,6 +255,41 @@ def check(arr, ind):
...
+        def np_new_axis_getitem(a, idx):
+            return a[idx]
+
+        def np_new_axis_setitem(a, idx, item):
+            a[idx] = item
+            return a
+
+        a = np.arange(4*5*6*7).reshape((4, 5, 6, 7))
+        idx_cases = [
+            (np.newaxis, slice(None)),
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:269]: As above, perhaps add some cases where `np.newaxis` isn't the first item?

File: numba/np/arrayobj.py [hunk 5]
```diff
@@ -981,7 +995,11 @@ def prepare(self):
         for i in self.indexers:
             i.prepare()
         # Compute the resulting shape
-        self.indexers_shape = sum([i.get_shape() for i in self.indexers], ())
+        shape_list = [i.get_shape() for i in self.indexers]
+        one = self.context.get_constant(types.intp, 1)
+        for i in self.newaxes:
+            shape_list.insert(i, (one,))
```
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1001]: This seems to be inserting at indexes based on a previous loop index into a list which has changing length. I guess this could work but it's quite hard to follow. This made me wonder if there's merit in implementing a `NewAxisIndexer` class as a way to handle the `newaxis` without there being a custom path and custom tracking/fixing of the shapes etc?

File: numba/tests/test_array_manipulation.py [hunk 6]
```diff
@@ -633,23 +629,40 @@ def test_squeeze_array_npm(self):
...
-        a = np.arange(9).reshape(3, 3)
+    def test_add_axis(self):
+        def np_new_axis_getitem(a, idx):
+            return a[idx]
+
+        def np_new_axis_setitem(a, idx, item):
+            a[idx] = item
+            return a
+
+        a = np.arange(4*5*6*7).reshape((4, 5, 6, 7))
```
Comment (stuartarchibald) [inline numba/tests/test_array_manipulation.py:640]:
```suggestion
        a = np.arange(4 * 5 * 6 * 7).reshape((4, 5, 6, 7))
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @kc611, I've left a few comments inline to look at, the most important of which is RE a potential new `NewAxisIndexer` class as an alternative implementation method, would be interested in hearing your thoughts. Thanks again!

File: numba/np/arrayobj.py (hunk 5 above)
Comment (kc611) [reply numba/np/arrayobj.py:1001]: I tried out this approach. The problem with it is, the way `Indexer`s built, they are supposed to provide an index for every axis in the 'source' array and not the resulting array. The number of `Indexer`s are supposed to correspond to the number of axes in the 'source'. (This is also represented by the following, which asserts that number of `Indexer` objects should always match the original dimensions of array.)

https://github.com/numba/numba/blob/bfe1270e23f8daf7fd95cfb6ef385f6c832fd927/numba/np/arrayobj.py#L1005

By adding a `np.newaxis`/`None`, we are adding a dimension in the resulting array and not the original one from where we're accessing the elements from. The original remains unchanged. 

So adding a `NewAxisIndexer`will leads to breaking of this `assert`s and requires special case handling in the index provision logic where we'll need to exclude the `NewAxisIndexer` objects since they don't have a existence in original array.

https://github.com/numba/numba/blob/bfe1270e23f8daf7fd95cfb6ef385f6c832fd927/numba/np/arrayobj.py#L1059

So it's possible, but will only add more complexity. Also we would want to take into consideration that currently the logic for both basic and advanced indexing is nearly the same. The above `Indexer`s will only apply to advanced indexing and not basic one. I think instead, we can somehow work on simplifying the current logic to make it less bug-prone/readable.

File: numba/tests/test_fancy_indexing.py (hunk 4 above)
Comment (kc611) [reply numba/tests/test_fancy_indexing.py:272]: Addressed in 89ba549

File: numba/tests/test_array_manipulation.py (hunk 3 above)
Comment (kc611) [reply numba/tests/test_array_manipulation.py:645]: Addressed in 89ba549

File: numba/tests/test_fancy_indexing.py [hunk 7]
```diff
@@ -255,6 +255,41 @@ def check(arr, ind):
...
+            (np.newaxis, slice(1), np.newaxis, np.array([1,2,1])),
+            (np.newaxis, slice(1), Ellipsis, np.newaxis, np.array([1,2,1])),
+            (np.newaxis, np.array([1,2,1]), np.newaxis, Ellipsis),
+            (np.newaxis, np.array([1,2,1]), Ellipsis, None),
+        ]
+        pyfunc_getitem = np_new_axis_getitem
+        cfunc_getitem = njit(pyfunc_getitem)
+
+        pyfunc_setitem = np_new_axis_setitem
+        cfunc_setitem = njit(pyfunc_setitem)
```
[the remark this answers is not in the document]
Comment (kc611) [reply numba/tests/test_fancy_indexing.py]: Addressed in 89ba549

File: numba/np/arrayobj.py (hunk 5 above)
Comment (stuartarchibald) [reply numba/np/arrayobj.py:1001]: Ah ok, thanks for considering it. I would expect `assert`s to need updating if logic changes, so I am not convinced that is an issue in itself, however additional/increased complexity is a problem and something to avoid. I think perhaps making the current logic have a more clear intent seems like the best thing to do?

**Review by stuartarchibald — COMMENTED** [review]
Additional thought... please could you add a note to the NumPy reference docs to state that `np.newaxis`/use of `None` as an axis is now supported? Many thanks!

File: docs/source/reference/numpysupported.rst [hunk 8]
```diff
@@ -161,9 +161,11 @@ Array access
 ------------
 
 Arrays support normal iteration.  Full basic indexing and slicing is
-supported.  A subset of advanced indexing is also supported: only one
-advanced index is allowed, and it has to be a one-dimensional array
-(it can be combined with an arbitrary number of basic indices as well).
+supported along with passing ``None``/ ``np.newaxis`` as indices for
```
Comment (stuartarchibald) [inline docs/source/reference/numpysupported.rst:164]:
```suggestion
supported along with passing ``None`` / ``np.newaxis`` as indices for
```

File: numba/np/arrayobj.py [hunk 9]
```diff
@@ -1208,7 +1235,6 @@ def maybe_copy_source(context, builder, use_copy,
             builder.store(builder.load(src_ptr), dest_ptr)
 
     def src_getitem(source_indices):
-        assert len(source_indices) == srcty.ndim
```
Comment (stuartarchibald) [inline numba/np/arrayobj.py:1211]: Should this be checking something like `len(source_indices) == srcty.ndim + len(self.newaxes)` or is this sort of invariant now entirely lost?

Comment (kc611) [reply numba/np/arrayobj.py:1211]: I removed it because during that function call we do not have the information regarding the number of `newaxes`/`None`s, since that is being unpacked/processed while calling the `Indexer`'s constructor which happens much later after this call. So at this specific line, yeah we can't do it unless we unpack the indices specifically for this `assert`.

Comment (stuartarchibald) [reply numba/np/arrayobj.py:1211]: Ah I see, thanks for the explanation. I think it's ok to leave this as-is then.

File: numba/tests/test_array_manipulation.py [hunk 10]
```diff
@@ -633,23 +629,49 @@ def test_squeeze_array_npm(self):
...
+        for idx in idx_cases:
+            expected = pyfunc_getitem(a, idx)
+            got = cfunc_getitem(a, idx)
+            np.testing.assert_equal(expected, got)
+
+            a_empty = np.zeros_like(a)
+            item = a[idx]
+
+            expected = pyfunc_setitem(a_empty, idx, item)
+            got = cfunc_setitem(a_empty, idx, item)
```
Comment (stuartarchibald) [inline numba/tests/test_array_manipulation.py:673]:
```suggestion
            expected = pyfunc_setitem(a_empty.copy(), idx, item)
            got = cfunc_setitem(a_empty.copy(), idx, item)
```
else they are both doing a `setitem` on the same array so the first could work and the second not and it wouldn't be possible to tell?

File: numba/tests/test_fancy_indexing.py [hunk 11]
```diff
@@ -256,6 +256,50 @@ def check(arr, ind):
...
+        for idx in idx_cases:
+            expected = pyfunc_getitem(a, idx)
+            got = cfunc_getitem(a, idx)
+            np.testing.assert_equal(expected, got)
+
+            a_empty = np.zeros_like(a)
+            item = a[idx]
+
+            expected = pyfunc_setitem(a_empty, idx, item)
+            got = cfunc_setitem(a_empty, idx, item)
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:300]:
```suggestion
            expected = pyfunc_setitem(a_empty.copy(), idx, item)
            got = cfunc_setitem(a_empty.copy(), idx, item)
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates, I think there's one remaining issue in the testing (noted inline) else this looks good.

File: numba/tests/test_array_manipulation.py (hunk 10 above)
Comment (kc611) [reply numba/tests/test_array_manipulation.py:673]: Yeah, this is a mistake.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes, great to see this long standing feature request implemented!
